// twitch-wrapper-ts/out/apiTypes.js
"use strict";
//#region Twitch Response Base
Object.defineProperty(exports, "__esModule", { value: true });
class TwitchSuccessResponse {
}
exports.TwitchSuccessResponse = TwitchSuccessResponse;
//#endregion
//#region users Endpoint
/**
 * Represents a Twitch user.
 * If it has the user:read:email scope, email will be returned and vice versa.
 */
class User {
}
exports.User = User;
/**
 * A result object from /users call.
 */
class UserResult {
}
exports.UserResult = UserResult;
/**
 * Contains who followed another user at which time.
 */
class Follow {
}
exports.Follow = Follow;
/**
 * Pagination object which contains the cursor.
 */
class Pagination {
}
exports.Pagination = Pagination;
/**
 * Success Twitch API Response from /users/follows.
 */
class UserFollowResult {
}
exports.UserFollowResult = UserFollowResult;
/**
 * Game information by game ID or name.
 */
class Game {
}
exports.Game = Game;
class GameResponse {
}
exports.GameResponse = GameResponse;
/**
 * A stream data returned by streams Endpoint.
 */
class Stream {
}
exports.Stream = Stream;
/**
 * The response returned by streams Endpoint.
 */
class StreamResponse {
}
exports.StreamResponse = StreamResponse;
/**
 * A video got from videos Endpoint.
 */
class Video {
}
exports.Video = Video;
/**
 * Data got from videos Endpoint.
 */
class VideoResponse {
}
exports.VideoResponse = VideoResponse;
/**
 * Got via a post request to clips Endpoint, as a clip is created.
 */
class CreatedClip {
}
exports.CreatedClip = CreatedClip;
/**
 * Clip data got from clips Endpoint.
 */
class Clip {
}
exports.Clip = Clip;
//#endregion
